scipy/stats/_bootstrap.py

```diff
@@ -0,0 +1,393 @@
+              n_resamples=9999, method='basic', random_state=None):
+    r"""
+    Compute a two-sided bootstrap confidence interval of a statistic.
+
+    When `method` is ``'percentile'``, a bootstrap confidence interval is
+    computed according to the following procedure.
+
+    1. Resample the data: for each sample in `data` and for each of
+       `n_resamples`, take a random sample of the original sample
+       (with replacement) of the same size as the original sample.
+
+    2. Compute the bootstrap distribution of the statistic: for each set of
+       resamples, compute the test statistic.
+
+    3. Determine the confidence interval: find the interval of the bootstrap
+       distribution that is
+
+       - symmetric about the median and
+       - contains `confidence_level` of the resampled statistic values.
+
+    While the ``'percentile'`` method is the most intuitive, it is rarely
+    used in practice. Two more common methods are available, ``'basic'``
+    ('reverse percentile') and ``'BCa'`` ('bias-corrected and accelerated');
+    they differ in how step 3 is performed.
+
+    If the samples in `data` are  taken at random from their respective
+    distributions :math:`n` times, the confidence interval returned by
+    `bootstrap` will contain the true value of the statistic for those
+    distributions approximately `confidence_level`:math:`\, \times \, n` times.
+
+    Parameters
+    ----------
+    data : sequence of array-like
+         Each element of data is a sample from an underlying distribution.
+    statistic : callable
+        Statistic for which the confidence interval is to be calculated.
+        `statistic` must be a callable that accepts ``len(data)`` samples
+        as separate arguments and returns the resulting statistic.
+        `statistic` must also accept a keyword argument `axis` and be
+        vectorized to compute the statistic along the provided `axis`.
+    axis : int, optional
+        The axis of the samples in `data` along which the `statistic` is
+        calculated. The default is ``0``.
+    confidence_level : float, optional
+        The confidence level of the confidence interval.
+        The default is ``0.95``.
+    n_resamples : int, optional
+        The number of resamples performed to form the bootstrap distribution
+        of the statistic. The default is ``9999``.
+    method : str in {'percentile', 'basic', 'bca'}, optional
+        Whether to return the 'percentile' bootstrap confidence interval
+        (``'percentile'``), the 'reverse'or the bias-corrected and accelerated
+        bootstrap confidence interval (``'BCa'``). The default is ``'BCa'``.
+        Note that only ``'percentile'`` and ``'basic'`` support multi-sample
+        statistics at this time.
+    random_state : {None, int, `numpy.random.Generator`,
+                    `numpy.random.RandomState`}, optional
+
+        If `seed` is None (or `np.random`), the `numpy.random.RandomState`
```

Comment:
If the minimum NumPy is >= 1.17, shouldn't `Generator` be preferred?  Same for passing an in `seed`, so that the only way to get a `RandomState` is to explicitly pass one.

Reply:
As far as I know, the minimum NumPy version we support still doesn't have Generator. I could branch based on version - is that what you had in mind? This branching would need to be documented, too.

Reply:
I think it should be the same, so if the minimum NumPy is still 1.16 or less, then the current would be the correct. I just hate to see new code primarily relying on RandomState, but this is the cost of backward compatibility?

Reply:
This argument should be processed like all of the other `random_state` arguments that exist in our code. If you want to change how we process all of these arguments, you can do so all at once later (presumably after scipy depends on 1.17+).

Right now, the `random_state=None` semantics mean that people can do `np.random.seed(...)` once and just call these functions and get controlled pseudorandomness. There is code out there relying on these semantics to continue to work.

Reply:
Neutral here. The new QMC is free from `RandomState` because being a new submodule, there is less risk of people falling into a trap here. But it this is directly accessible from `scipy.stats` so it kinda makes sense to keep it compatible with 1.16.

But sure, once we branch of 1.7, the first thing I will do is a  big PR to remove all occurence of `RandomState`!

Reply:
> This argument should be processed like all of the other `random_state` arguments that exist in our code. If you want to change how we process all of these arguments, you can do so all at once later (presumably after scipy depends on 1.17+).

Then it could have a new, future proof name, so that it doesn't conflict with the `random_state` keyword. New keyword, new semantics.  TBH the use of `random_state` as a keyword, while completely reasonable, makes doing anything other than preferring a `RandomState` strange. Sounds like it might be moot anyway if compatibility with NumPy 1.16 is a requirement.

Reply:
The name is part of an on-going discussion over at https://github.com/scipy/scipy/pull/13955#issuecomment-830818617

Reply:
Adding a new argument with different semantics doesn't really change the main issue, which is what happens when the user calls the function without either argument. Right now, it means it uses the global `RandomState` in `np.random`, which can be controlled with `np.random.seed()`, as is done in these unit tests, for instance.

Eliminating that would be a very large change. It's not a no-brainer even after scipy depends on numpy 1.17+.

Reply:
I would suggest never introducing `random_state` as an argument in 2021 era code.  But I can understand if the new structure isn't quite settled yet, as @tupui indicates.

Reply:
> I would suggest never introducing `random_state` as an argument in 2021 era code.
Do you mean that we should have non-deterministic code? So something like hypothesis testing?

Reply:
> Eliminating that would be a very large change. It's not a no-brainer even after scipy depends on numpy 1.17+.

Do you mean that it will be hard because of all the tests to change? Or that we could break users codes? Because having something like the `check_random_state` from QMC accepts `RandomState` for testing purposes only. So IMO it would be fine.

Reply:
No, he means we should have a new argument (maybe named `seed`) that only deals with `Generator`.

Reply:
> Or that we could break users codes?

This is always the pain point. Internal stuff, even pervasive change, is always easy by comparison to external.

Reply:
Ah ok then yes, `seed` is what we went for with QMC.

Reply:
What is the conclusion here @tupui? As a reviewer who has indicated approval, what needs to be changed?

Reply:
For me both approaches are working. In the long run we probably want the generator version (so copying what is done in QMC and use it's `check_random_state`. But we agreed earlier with what you currently have. Thus for me the PR is still a go as it is and this should not block it. @rkern what do you say? Shall we directly go for the new way? I thought I remembered we said this should still use the old way.

Reply:
To be clear (sorry it's late and in rush), I am more than happy to go with the new interface and use `seed` instead. So this takes a +1 for me. I just thought we said we where not quite there yet. But if another maintainer validates the `seed`, then `seed` it is.

Reply:
I'd rather stick with this until we can make a wholesale change to all of `scipy.stats`.

Reply:
I interpret this as +2 for leaving this as-is (`random_state`) and changing everything together when ready. I appreciate it.